crates/esthri/src/ops/upload.rs

```diff
@@ -336,39 +335,22 @@ where
                 Ok(None)
             } else {
                 let read_size = u64::min(remaining, read_part_size);
-
                 let mut buf = BytesMut::with_capacity(read_size as usize);
                 buf.resize(read_size as usize, 0);
-
-                let reader = reader.unwrap();
-
-                let result: Result<Bytes> = {
-                    let reader = reader.clone();
-                    let mut reader = reader.lock().await;
-                    let reader = &mut *reader;
-
+                {
+                    let mut guard = reader.lock().await;
                     let seek = chunk_size * chunk_number + part_number * read_part_size;
-                    reader.seek(SeekFrom::Start(seek)).await?;
-
+                    guard.seek(SeekFrom::Start(seek)).await?;
                     let slice = buf.as_mut();
-                    reader.read_exact(&mut slice[..read_size as usize]).await?;
-
-                    let body = buf.freeze();
-                    Ok(body)
-                };
-
-                match result {
-                    Ok(buf) => {
-                        if read_size == 0 {
-                            Err(Error::ReadZero)
-                        } else {
-                            Ok(Some((
-                                buf,
-                                (remaining - read_size, part_number + 1, Some(reader)),
-                            )))
-                        }
-                    }
-                    Err(err) => Err(err),
+                    guard.read_exact(&mut slice[..read_size as usize]).await?;
+                }
```

Comment:
No need to clone the Arc here, as long as you drop the guard after reading from the file you can return the original reader. When you think about it this makes sense, each chunk is read serially, so we only need one reader. The reader cloning comes into play reading multiple chunks at a time